File: Stage27/Java/Problem16928.java
package Stage27.Java;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Problem16928 {
    static int[] board = new int[101];
    static boolean[] visited = new boolean[101];

    static int bfs() {
        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{1, 0});
        visited[1] = true;

        while(!queue.isEmpty()) {
            int[] cur = queue.poll();
            int position = cur[0];
            int count = cur[1];

            if (position == 100) {
                return count;
            }

            for (int dice = 1; dice <= 6; dice++) {
                int next = position + dice;

                if (next > 100) continue;

                if (board[next] != 0) {
                    next = board[next];
                }

                if (!visited[next]) {
                    visited[next] = true;
                    queue.offer(new int[]{next, count + 1});
                }
            }
        }
        return -1;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            int from = Integer.parseInt(st.nextToken());
            int to = Integer.parseInt(st.nextToken());
            board[from] = to;
        }

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int from = Integer.parseInt(st.nextToken());
            int to = Integer.parseInt(st.nextToken());
            board[from] = to;
        }

        System.out.println(bfs());
        br.close();
    }
}

File: Stage27/Python/Problem16928.py
import sys
from collections import deque

input = sys.stdin.readline

def bfs(board):
    visited = [False] * 101
    queue = deque()
    queue.append((1, 0))
    visited[1] = True
    
    while queue:
        position, count = queue.popleft()
        
        if position == 100:
            return count
        
        for dice in range(1, 7):
            next_pos = position + dice
            
            if next_pos > 100:
                continue
            
            if board[next_pos] != 0:
                next_pos = board[next_pos]
                
            if not visited[next_pos]:
                visited[next_pos] = True
                queue.append((next_pos, count + 1))
    return -1

def main():
    N, M = map(int, input().split())
    board = [0] * 101

    for _ in range(N):
        a, b = map(int, input().split())
        board[a] = b

    for _ in range(M):
        a, b = map(int, input().split())
        board[a] = b

    print(bfs(board))
    
if __name__ == "__main__":
    main()